### DISCUSSION TITLE
Discussion on Reusing Templates

### DISCUSSION THREAD

**Discussion by hminghe** [vuejs/rfcs#517] [General]
### What problem does this feature solve?

- I don't want to use jsx

For example:
```vue
<script lang="ts" setup>
defineProps<{
  title: string
  unit: string
  unitIsRight: boolean
}>()
</script>

<!-- reuse div.unit -->
<template>
  <div>
    <div v-if="!unitIsRight" class="unit">
      <slot name="unit">
        {{ unit }}
      </slot>
    </div>

    <div>{{ title }}</div>

    <div v-if="unitIsRight" class="unit">
      <slot name="unit">
        {{ unit }}
      </slot>
    </div>
  </div>
</template>

```

写组件的时候经常会有这个问题，然后又换成 `tsx` 写，但是 `tsx` 的 `slot` 和 `expose` 是无法实现代码提示的。
You often get this problem when you write components, and then switch to `tsx`, but  `slot` and `expose` don't allow for code hints.

### What does the proposed API look like?

```vue
<script lang="ts" setup>
 defineProps<{
  title: string
  unit: string
  unitIsRight: boolean
}>()
</script>

<template name="unit">
  <div v-if="!unitIsRight" class="unit">
    <slot name="unit">
      {{ unit }}
    </slot>
  </div>
</template>

<template>
  <div>
    <template v-if="!unitIsRight" is="unit"></template>

    <div>{{ title }}</div>

    <template v-if="unitIsRight" is="unit"></template>
  </div>
</template>
```

Template to: 
```js
return (_ctx, _cache) => {
    const __template = {
      unit: () => (_openBlock(), _createElementBlock('div', { class: 'unit' }, [
        _renderSlot(_ctx.$slots, 'unit', {}, () => [
          _createTextVNode(_toDisplayString(props.unit), 1 /* TEXT */),
        ]),
      ])),
    }

    return (_openBlock(), _createElementBlock('div', null, [
      (!props.unitIsRight)
        ? __template.unit()
        : _createCommentVNode('v-if', true),
      _createElementVNode('div', null, _toDisplayString(props.title), 1 /* TEXT */),
      (props.unitIsRight)
        ? __template.unit()
        : _createCommentVNode('v-if', true),
    ]))
  }
```

**Comment by Rolanddoda** [comment]
Interesting! I always go for creating a child component when I see repetition and while reading the problem I intuitively thought of solving it that way.

But having multiple templates in order to reuse them inside a SFC looks amazing. I am thinking that this way of doing it, wouldn't require a new instance to be created right? Because if a new instance is created, then I guess there is no need to do that, we could just create a child component.

So in terms of optimizations, having this way of reusing templates, would probably result in fewer component instances in userland.

Thinking of logical concerns, this can align well I believe. Think for example:

```js
<script setup>
const count = ref(0)

function increase() {
  count.value++
}

const user = ref({ name: 'Roland' })

function clear() {
 user.value = { name: '' }
}
</script>

<template name="count">
 Count is: {{ count }}
 <button @click="increase">Increase</button>
</template>

<template name="user">
 User name is: {{ user.name }}
 <button @click="clear">Clear</button>
</template>

<template>
 <div>
   <template is="count"></template>
   <hr />
   <template is="user"></template>
 </div>
</template>
```

**Reply by louiss0** [reply]
While I don't mind using jsx or components. This seems like a good idea I was thinking about a similar api they are called blocks. A block is just a template just like any other but then you could reference it using component.

**Comment by sxzz** [comment]
~~🚧 WIP in Vue Macros: https://github.com/sxzz/unplugin-vue-macros/pull/111~~

---

🎉 Beta version released as `v0.14.0-beta.0`. https://beta.vue-macros.sxzz.moe/features/named-template
It's a very early stage version, please help to test and contribute more usage cases. ❤️

**Reply by sxzz** [reply]
🎉 Beta version released as `v0.14.0-beta.0`. https://beta.vue-macros.sxzz.moe/features/named-template
It's a very early stage version, please help to test and contribute more usage cases. ❤️

**Reply by HassanZahirnia** [reply]
Hey I have a question,
I wonder if it would be possible to use this approach to create "mini components" inside SFCs ? So you don't go create a separate SFC just for a small abstraction.
Say when for example you have a section in your page where it gets repeated like 20 times, but it only happens in that specific page. So you create a small template/script inside SFC and use that inside the main template, inside of going and creating a separate SFC just to abstract something in a single SFC. Hope that makes sense.

**Reply by Rolanddoda** [reply]
@HassanZahirnia this is exactly what this is about. Reusing templates without needing to create a new instance/sfc component

**Reply by HassanZahirnia** [reply]
@Rolanddoda Yeah but @sxzz 's implementation shows a different scenario/use case. I was not sure tbh so I asked to be sure

**Reply by Rolanddoda** [reply]
@HassanZahirnia yeah always good to ask. Can you please elaborate on how the scenario is different? From what I see, you can have multiple templates and reuse them inside the "main" template. One thing that I don't see in the docs of that link above, is the ability to have slots or any kind of dynamic behavior. Everything looks static. @sxzz is it in the plan to add dynamic behavior by slots?

**Reply by HassanZahirnia** [reply]
Yeah you're right! Tbh the example used to introduce this feature was a bit unusual for me 😅
Honestly I think the syntax/naming could be better, right now it's very confusing. Like first glance you could not tell which template is the "main" one.

Maybe you could not use `template` for the named templates ? Rename it to something like :
```html
<miniTemplate name="title">
  <div class="text-xl">
    <slot />
  </div>
</miniTemplate>
```
This way it's distinguishable from the main `<template>` in the SFC.

**Reply by Rolanddoda** [reply]
@HassanZahirnia yeah the syntax will be bloated with many `template` tags and it won't look nice, I agree.

If I could suggest a better syntax, this is what comes to my mind right now:

```html
<script setup>
const count = ref(0)

function increase() {
  count.value++
}

const user = ref({ name: 'Roland' })

function clear() {
 user.value = { name: '' }
}
</script>

<unit name="count">
 Count is: {{ count }}
 <button @click="increase">Increase</button>
</unit>

<unit name="user">
 User name is: {{ user.name }}
 <button @click="clear">Clear</button>
</unit>

<template>
 <div>
   <unit is="count"></unit>
   <hr />
   <unit is="user"></unit>
 </div>
</template>
```

Haven't added something for `slot` but it would be nice if we can have the ability to use them and that would truly unlock new possibilities and in userland, there will be fewer instances.

@sxzz thoughts?

**Reply by sxzz** [reply]
I think using `<unit is` may cause a breaking change when you have a component called `unit`. And `<template>` is pretty good and I can't see any benefits about using another name for the child template, in my opinion. So maybe we could have an RFC to discuss this feature further (in vuejs/rfcs).

> is it in the plan to add dynamic behavior by slots?

I think static behavior is enough for now, so I may implement the dynamic version in the future, or PR welcomed.

**Comment by eneajaho** [comment]
For this feature to really shine, you would have to give the possibility to pass context to that template. 

What do I mean by this? 

For example, you have to show 2-3 different templates inside a v-for, but you have to show them in different conditions and with different data structures. 

If we can have this, it means that we can also have the possibility to pass around this templates in order to be reused in other components. 


```vue
<script setup>
const users = ref([ 
 {name: 'a', lastname: 1, onlyName: true},
 {name: 'b', lastname: 2, onlyName: true},
 {name: 'b', lastname: 3, onlyName: false},
]);
</script>

<v-template #userCard1 let-user>
   User name is: {{ user.name }}
</v-template >

<v-template #userCard2 let-user>
   User last name is: {{ user.lastName }}
</v-template>

<template>
 <div v-for="user in users">
    <v-template-outlet v-if="user.onlyName" :template="userCard1" :context="user"></v-template-outlet>
    <v-template-outlet v-if="!user.onlyName" :template="userCard2" :context="user"></v-template-outlet>
 </div>
</template>
```

The idea is taken from Angular [(ng-template, ng-template-outlet )](https://angular.io/api/common/NgTemplateOutlet)

**Reply by louiss0** [reply]
I like this idea It's simply additive and it does not destroy any other mechanisms of Vue. However, I'd change the `#userCard1` or 2 to just name and `let-user` to just context.

**Comment by hminghe** [comment]
# Reference slots
```vue
<script setup>
const users = ref([ 
 {name: 'a', lastname: 1, onlyName: true},
 {name: 'b', lastname: 2, onlyName: true},
 {name: 'b', lastname: 3, onlyName: false},
]);
</script>

<template #userCard1="ctx">
   User name is: {{ ctx.user.name }} {{ ctx.test }}
</template >

<template #userCard2="{ user, test }">
   User name is: {{ user.name }} {{ test}}
</template>

<template>
 <div v-for="user in users">
    <template is="userCard1" :user="user" test="1"></template>
    <template is="userCard2" :user="user" test="2"></template>
 </div>
</template>
```

**Reply by rubnvp** [reply]
@hminghe that's exactly what I was thinking too 👌, looks great! I would even include the ability to pass slots inside `<template>` 👍

**Reply by a1mersnow** [reply]
good design. But this means we need to add one more meaning to `v-slot`, right?

**Comment by louiss0** [comment]
# Template Block Proposal

When It comes to this proposal. I'd like to say that even though JSX is just better to use I do support this proposal. I think there are two ways to implement this feature. I call this Open vs Closed blocks. The way that I want this proposal to be implemented is simple. We get two new components. A block component and a outlet component. The reason why people have made this proposal is for reasons related to using JSX. 


## Weaknesses of JSX 

1. You gain no autocomplete when using slots.
2. You have no access to directives at all 
3. You have to think in two different ways when it comes to rendering components   


## Vue Block

The `vue:block` is the component that has the html written inside of it. The point of this component is to allow the user to write html that is to be rendered by another component called a `vue:block-outlet` . A block is a template that does nothing at all unless it is rendered by an outlet. This kind of component can either have access to the state or not have any access to the state at all. If it does then they are **open-blocks**. If not then they are **closed-blocks**. A closed block must take in a prop and use that prop to render dynamic data.     

### Open Block

**This component has one prop**

- `name` the name that the `vue:block-outlet` to use render it's self 

**Creation of Open Block Vue block**

```vue
<script setup>

defineProps<{name:string}>()

</script>

// This template works like a normal template 
// It is not picked up by vue
<template v-bind="{name}">
  <slot/>
</template>

```


**Implementation of Vue block **

```vue

<script setup>

const foo = "foo"

</script>  

<vue:block name="random-text">

// The content in here will not be rendered 
unless the outlet is used 

{{ foo }}

</vue:block>

```

### Closed Block

**This component has two props**

- `name` the name that the `vue:block-outlet` to use render it's self 
- `context` the data that is to be passed to the block by the outlet

**Creation of Vue Closed Block**

```vue
<script setup generic="T extends Record<string,unknown>">

defineProps<{name:string, context:T}>()

</script>

// This template works like a normal template 
// It is not picked up by vue
<template v-bind="{name}">
  <slot name="context" v-bind="{...context}"  />
</template>

```

**Implementation of Vue block**
```vue
<script setup>

const foo = "foo"

</script>  

<vue:block name="random-text" #context="{foo}">

// The content in here will not be rendered 
unless the outlet is used 

{{ foo }}

This has no acesss to the state so we grab to through the context. 

</vue:block>

```

## Vue Block Outlet

A block outlet is the component that decides where a  `vue:block` component should be rendered. It's a component that allows the user to render the contents of a block in a certain area by simply using it's name. What it does is inject content that is inside of a `vue:block` into itself. If block are meant to be open then it just just needs the block prop as it's only prop. If blocks are closed then it only needs the `context` prop and the `block` prop. It then passes that into the `vue:block` who's content is being rendered.     


**Implementation of Block Outlet **

```vue
// look at the block example to find out where these props come from
<vue:block-outlet block="random-text" :context="{foo}">

```


## Examples

Like I said in previous examples. I talked about the difference between closed and open blocks. An open block automatically has access to the state of the app. A closed block doesn't. I will demonstrate how one would work vs the other. 

**Let's say that this is my Vue component.**  

```vue
<script setup>

const links = [
"Home",
"About",
"Contact",
"Services",
]

</script>

<template>
	<nav>
		// this is supposed to be a hamburger menu
		<div>
		<span></span>
		<span></span>
		<span></span>
		</div>
		// this is a set of links
		<ul>
		<li v-for="link of links">
		{{link}}
		</li>	
		</ul>
		
		</nav>
</template>

```

**My Vue component using open blocks**

```vue

<script setup>

const links = [
"Home",
"About",
"Contact",
"Services",
]

</script>

<template>
	<nav>
		// this is supposed to be a hamburger menu
		<div>
		<vue:block-outlet block="hamburger-menu" />	
		</div>
		// this is a set of links
		<ul>
		<vue:block-outlet block="pages-links" />	
		</ul>
		
		</nav>
</template>

<vue:block name="hamburger-menu">
<span></span>
<span></span>
<span></span>
</vue:block>	

<vue:block name="pages-links">
<li v-for="link of links">
	{{link}}
</li>
</vue:block>


```


**My Vue component using closed blocks **

```vue

<script setup>

const links = [
"Home",
"About",
"Contact",
"Services",
]

</script>

<template>
	<nav>
		// this is supposed to be a hamburger menu
		<div>
		<vue:block-outlet block="hamburger-menu" />	
		</div>
		// this is a set of links
		<ul>
		<vue:block-outlet 
		block="pages-links" 
		:context="{links}"
		/>	
		</ul>
		
		</nav>
</template>

// These blocks have no acess to component state
<vue:block name="hamburger-menu">
<span></span>
<span></span>
<span></span>
</vue:block>	

<vue:block name="pages-links" #context="{links}" >
	<li v-for="link of links">
		{{link}}
	</li>
</vue:block>

```

**Reply by a1mersnow** [reply]
Not concise

**Reply by louiss0** [reply]
It's hard to explain the outlet part of the block template. I hope you understand the point. The blocks are the parts of the template that are hidden. The outlets are the components that render the content that is inside the blocks. If you would like to tell me which part of the syntax is not concise. I would like to fix it.

**Reply by louiss0** [reply]
By the way If you are confused about the Closed vs Open Block Thing. 
- A closed block is a block that has no access to the state of the component at all
- An open block is a block that has access to the state.  

This proposal talks about two different ideas on how to do the same thing.

**Reply by a1mersnow** [reply]
I understand the open block and close block.

Open block: [The original post](https://github.com/vuejs/core/discussions/6898#discussion-4480034)
Close block: [This comment](https://github.com/vuejs/core/discussions/6898#discussioncomment-3905380)

I just think the names `vue:block` and `vue:block-outlet` are not concise……

**Reply by a1mersnow** [reply]
I think it's better to just use `<template name=...>` and `<template is...>`

And I prefer open block because it's easy to use……
If I have time to write a close block, why not write a new component.

**Reply by a1mersnow** [reply]
Open block share the same context with main template, while close block needs to create a new context which will cause performance cost that I think is't worth to pay.

Open block is just like copy the text in template and paste into the dest location.

**Reply by louiss0** [reply]
Good then I'd like to say that since you like the idea of an is component. There is a component that does use `is` already. It's called component.

**Reply by louiss0** [reply]
The reason decided on a new api is because vue needs the template tag. It uses it to render components. I don't know why it does. but that just seems to be the case. I didn't want ask the team to destroy an aspect of the framework that allows it to work.

**Reply by louiss0** [reply]
Hopefully one day I don't have to write template just to write a simple. Button component. Because for simple component's template was just overhead but now it can be used for important things if they decide to make it so that you don't have to write it just to create a component

**Comment by CarelessCourage** [comment]
I think this is a great idea. How would we import these though? What I really want is to be able to use this approach to make multiple components that share state and or style. Like for instance, a togglable dropdown where you can import the dropdown and the toggle handle seperatley and place them however you want while still having the toggle be able to trivially manipulate the state of the dropdown. Example:
 
```vue
<script setup>
const dropdown= ref(false)
</script>

<template name="handle">
 <button @click="dropdown = !dropdown">{{dropdown ? "close" : "open"}}</button>
</template>

<template name="dropdown">
 <div :class="dropdown"  :class="{ open: dropdown }">
    <slot/>
 </div>
</template>

<style scoped>
.dropdown {
  overflow: hidden;
  padding: 12px;
  max-height: 300px;
  transition: .4s;
}

.dropdown:not(.open) {
  max-height: 0px;
}
</style>
```

and then import like this: 

```vue
<script setup>
import Shelf from "./components/shelf.vue"
</script>

<template>
 <Shelf.handle/>
 <div>Some other random stuff to prove a point</div>
 <Shelf.dropdown>
   <p>Some item 1</p>
   <p>Some item 2</p>
   <p>Some item 3</p>
   <p>Some item 4</p>
 </Shelf.dropdown>
</template>
```

I feel like these kinds of DX wins could help give Vue an edge over things like Svelte in particular. Since svelte doesnt encapsulate its html in template tags but Vue does. Vue has a uniqe opportunity to extend this already existing pattern to compose multiple templates in ways that Svelte would require a bigger deviation from its established patterns to be able to match.

**Reply by louiss0** [reply]
I feel like you are asking for a bit much but since this feature is a feature that people want. I guess I'd like to ask what if I would not want a template to be exposed to the outside?

**Reply by louiss0** [reply]
The JSX hate is real. I hope this works out. I don't have to switch contexts and I don't miss out on directives.

**Reply by CarelessCourage** [reply]
Maybe they only get exposed if you give them a name attribute? Im not sure I understand the question fully though. Importing a componennt like this: "shelf.dropdown" would be the same as just importing Shelf regularly. In either case you are importing the template and the state same as you always have. Its just that - if you have multiple templates in one file, you can now decide to import the component with just one of the templates attached. If you dont want to expose that then just dont make multiple templates in your file - or dont import them if you do. Or, as I said, if this is really a problem then just decide if the template is pickable depending on whether it has a name attribute to be picked from in the first place.

**Comment by qmhc** [comment]
It very useful idea, it can maintain consistency and avoid jsx syntax.

I would like to let them as `<part>` tag, and using `v-part` to use them.

```vue
<script setup>
import { ref } from 'vue'

const msg = ref('msg')
const count = ref(0)

function inc() {
  count.value++
}
</script>

<part name="one">
  <span>{{ message }}</span>
</part>

<part name="two">
  <span @click="inc">{{ count }}</span>
</part>

<template>
  <div>
    <template v-part:one></template>
    <template v-part:two></template>
  </div>
</template>
```

**Comment by jd-solanki** [comment]
After refactoring my lib [Anu](https://github.com/jd-solanki/anu)'s code from TSX to SFC, This is the only missing feature in SFC over TSX.

In TSX I was assigning the btn content to some variable and later wrap the loader based on the loading prop. This was super easy in TSX however with SFC and template tag there's no other way.

Thanks @sxzz I don't have to refactor the button component from SFC to TSX again 🤣 

I'm using named template from [vue-macros](https://vue-macros.sxzz.moe/features/named-template.html) for now.

**Comment by antfu** [comment]
Found an approach that requires no tooling support and you can use today:
https://github.com/antfu/vue-reuse-template

It also supports complete type checks. Give it a try, and I'd love to hear your feedback :)

**Reply by hminghe** [reply]
真没想到这个方式。
简单方便，编辑器和TS都能支持了。

**Reply by qmhc** [reply]
Nice job. Might it be added to native support?

**Reply by Rolanddoda** [reply]
That's very smart! Would love to see how it works performance-wise and a section in the readme on how it works! Really impressive!

**Reply by antfu** [reply]
There is almost zero performance cost, I believe performance is not a concern using it.

**Reply by jods4** [reply]
This is slick and useful!

I hope the idea can mature and be adopted into core? It seems worthy of that promotion.

Inside core, `<DefineTemplate>` could be an independent `<template name=abc>` block, that seems cleaner. 

And then reuse could be through a new directive that triggers `<template>` special handling (same as `v-if` or `v-for` already do today): `<ReuseTemplate>` would be `<template v-named=abc>`.

Implementation works same as your lib. That would be cool DX out of the box!

**Reply by dolanske** [reply]
Would love to see this in core, super useful

**Reply by a1mersnow** [reply]
@jobs4 I'm uncertain if incorporating this into the core is advisable for several reasons. Firstly, although the syntax you suggest is more concise, it necessitates an extra compilation step. Secondly, as Vue is a progressive framework, it poses a challenge to construct such a component without the support of Single-File Components. One potential solution could be introducing a new interface named namedTemplates. Lastly, we must exercise caution when adding new directives to ensure stability and scalability.

**Reply by antfu** [reply]
To make it easier, it's now part of VueUse:
https://vueuse.org/core/createReusableTemplate/

**Reply by jods4** [reply]
@a1mersnow

1. "An extra compilation step"
Vue compiles templates already, this would not be a whole new toolchain step, merely an extension of the existing syntax / compiler.

2. Challenges to construct such a component outside SFC
Do you even _need_ that outside SFC??
If you're writing rendering functions, you'd simply use a local function for re-usable parts, there's no need for this.
I'm not an expert in JSX as I don't use it myself, but I expect it would be the same: a local function that produce a JSX fragment achieves this, doesn't it?
So this is a perfect fit for being a compiler transform: in the same sense that `v-if` or `v-for` are SFC/templates/compiler only, you use JS `if`, `for` to achieve the same outside compiled templates. Reusable template parts in SFC translate into JS functions.

3. Caution when adding features
True, and I leave this to the judgement of core team.
I feel like this is nice and generally useful, IMHO worth of being part of the core template syntax/features.
The fact that it can be emulated in a user-land library is a good reason to question this and consider keeping it outside, I agree.

**Reply by louiss0** [reply]
> Found an approach that requires no tooling support and you can use today: https://github.com/antfu/vue-reuse-template
> 
> It also supports complete type checks. Give it a try, and I'd love to hear your feedback :)

There is a new plugin called [Vite Plugin Nested SFC](https://www.npmjs.com/package/vite-plugin-vue-nested-sfc) 
It solves all of our problems this discussion is simply over please rejoice and **PUSH TO COOORRREE**

**Reply by louiss0** [reply]
> Would love to see this in core, super useful

There is a new plugin called [Vite Plugin Nested SFC](https://www.npmjs.com/package/vite-plugin-vue-nested-sfc) 
It solves all of our problems this discussion is simply over please rejoice and **PUSH TO COOORRREE**

**Reply by louiss0** [reply]
> This is slick and useful!
> 
> I hope the idea can mature and be adopted into core? It seems worthy of that promotion.
> 
> Inside core, `<DefineTemplate>` could be an independent `<template name=abc>` block, that seems cleaner.
> 
> And then reuse could be through a new directive that triggers `<template>` special handling (same as `v-if` or `v-for` already do today): `<ReuseTemplate>` would be `<template v-named=abc>`.
> 
> Implementation works same as your lib. That would be cool DX out of the box!

There is a new plugin called [Vite Plugin Nested SFC](https://www.npmjs.com/package/vite-plugin-vue-nested-sfc) 
It solves all of our problems this discussion is simply over please rejoice and **PUSH TO COOORRREE**

**Reply by texastoland** [reply]
The composable (and previously macro) is amazing 🙌🏼 I hope something gets considered for core nonetheless. A minimal `<template>`-only syntax would improve DX and discoverability for a common use case.

**Reply by louiss0** [reply]
Did you try [Vite Plugin Nested SFC](https://www.npmjs.com/package/vite-plugin-vue-nested-sfc) I think you may like it better.

**Reply by texastoland** [reply]
@louiss0 You spammed every single reply in this issue. 1 was enough. It's unhelpful 🤫

**Reply by louiss0** [reply]
I did that so that each person could get notified as you did. I think they will see it. The problem is that no one is talking about that plugin.

**Reply by texastoland** [reply]
Nobody in any community wants to be spammed. `PUSH TO COOORRREE` isn't part of the RFC process either. Please stahp 🙏🏼

**Reply by louiss0** [reply]
Fine I just wanted to put an idea out there. I have to find a different way for people to notice what I say on the internet. I got excited when I saw that plugin.

**Reply by louiss0** [reply]
I'm sorry I bothered you I decided to start a new discussion on this topic. vuejs/core#8651. I put my thoughts out there in full. I'm not going to bother you anymore. But I think you should consider what I was talking about. This is an important issue.

**Comment by louiss0** [comment]
@hminghe Thank you for bringing up this topic this is a problem that needs to be solved the optimal way of doing this is to use nested SFC's See an example [here](https://stackblitz.com/edit/vitejs-vite-iroyak?file=src%2Fcomponents%2FHelloWorld.vue). It's called vite-plugin-vue-nested-sfc

**Reply by texastoland** [reply]
I'm interested what you like about https://github.com/HunYan-io/vite-plugin-vue-nested-sfc compared to other alternatives? I'm going to look at vuejs/core#8651 later. I agree it'd be nice to have something in core. But I'm not sure it aligns with the core team's priorites.

1. I'm using the [official VueUse composable](https://twitter.com/texastoland/status/1672843009862270977?s=20) announced in https://github.com/vuejs/core/discussions/6898#discussioncomment-5614265. It accomplishes everything I need except `v-model` binding. But that's part of a bigger issue in #5899.
2. The macro from popular Vue Macros announced in https://github.com/vuejs/core/discussions/6898#discussioncomment-3896329 didn't support props or Volar and is now deprecated.
3. https://github.com/liulinboyi/unplugin-vue-reuse-template for Vite shares similar syntax and limitations with the macro.

**Reply by louiss0** [reply]
I think this one is the best because it solves all of our problems. The only problem with it so far is that it's a hack and not a native solution.
I know that the recommended solution is nice but sometimes you want to emit events or use state. This plugin allows you to create multiple components in a file. These components work like regular components. 

There are some costs of using this plugin so far. 

- It does not work with SSR.
- It's tied to vite.
- You can't just put multple components in a file without having to write an empty script or file tag.
 
 Those reasons are why I'm advocating for this solution.

Other than the costs above this solution is superior to the other ones because you are just writing multiple components in one file. 


```html

<template>
  <h1>Home</h1>
<HelloWorld />
 <Counter />
</template>

<component lang="vue" name="hello-world">
  <template>
    Hello World 
  </template>
</component>

<component name="counter">
  <script lang="ts" setup>
  
  import { ref } from 'vue';
  
  const count = ref(0);
  
  </script>

  <template>
    <button type="button" @click="count++">count is {{ count }}</button>
  </template>
</component>



```

**Reply by texastoland** [reply]
I linked another Vite plugin above. I agree the 1 you're using is unique because it wraps an entire SFC in a tag. For me the most important feature is integration with other tools. VueUse is the only option with Volar support because it's "just Vue". It's also the most likely to be maintained.

SFCs conceptually enforce code organization. The use case for additional components would be to reduce duplication such as boilerplate like form elements or class names. In theory a nested component shouldn't need `emit` because you can pass callbacks as props.

Nested SFCs and vuejs/core#8651 bypass the _encapsulation_ of SFCs in general. When they need state it's probably time for a new SFC anyway. Thanks for creating a dedicated thread to share 💯

**Comment by ShenQingchuan** [comment]
Hi, guys! There's another approach coming out named "Vue Vine", and you can easily write multiple components, even just contain a bare template. Check out more details on [Vue Vine docs](https://vue-vine.dev).

I hold the opinion that putting multiple templates (or we can say it is small fragments) is breaking the basic concept of "Single" file component.

Over many years of iteration, the boundaries of SFC components are always first set in a single file, it doesn't make sense to create more rugged and winding approach to overpass its basic concepts, we should look for another way instead.

So, I'm providing a more flexible organization approach for you to be able to write multiple components in 1 TS file, just like this:

![image](https://github.com/vuejs/rfcs/assets/46062972/7442006e-175d-4c27-94a5-a60f8efbcd07)

**Comment by mary-ext** [comment]
with Svelte introducing this sort of reusable templating functionality in the form of snippet blocks I wonder if we can revisit this now, it'd be very nice to have this functionality in Vue. We've had several components where we've had to duplicate our templating for one reason or another, where having them as separate components doesn't seem worth doing, so having this would be a big plus.

**Reply by rudolf61** [reply]
I totally agree. I have recently started using Svelte 5. Snippets transform UI composition from a "specialized container" pattern (Slots) into a functional pattern.

In Vue or Svelte 4, slots require specific tags (<slot /> or <template #name>). Snippets, however, use standard logic. Passing a snippet is syntactically identical to passing a variable or a function, which is much simpler to comprehend, for me at least :). While Vue has "Scoped Slots," snippets make prop-passing feel more like a standard function call. This makes complex scenarios—like rendering a custom row in a data table—much more intuitive.

Unlike a slot, which is "consumed" by a specific location in a child component, a snippet is a reusable block of logic. It can be defined once and invoked multiple times within the same file or passed across multiple components.

Although it is possible to use slots in Svelte 5, I have understood that these are compiled into snippets. I sincerely hope that Vue will introduce snippets.

**Comment by theoephraim** [comment]
Svelte snippets now support [declaration tags](https://svelte.dev/docs/svelte/declaration-tags) - meaning they can have their own props and state.

I do think building something here could be very useful to avoid creating tiny simple components. I don't think embedding multiple full SFCs in the same file makes sense as it would require huge tooling changes. But I do think some kind of _limited_ inline template/snippet could make sense. We could call them "snippets" too if we want to differentiate from full components, although I think having multiple "template" tags works too.

I think you want to use an SFC block, and you can use attributes on that block to define props and state. Then you can use it within the template like a normal component. Have not thought through all the specific implementation/syntax, but something like
```vue
<snippet name="like-button" props="{label: String}" state="{ likeCount: 0 }">
```
or 
```
<snippet name="like-button" label:prop="" likeCount:state="0">
```
Obviously lots of options there and need to figure out how to best get flexibility with TS support...


Could also declare the logic in the main script block
```
<script setup>
defineSnippet</* props shape */>('foo', () => {
  // define state and internal behaviour
})
</setup>
<snippet name="foo">...
```

or use another script block tagged with the snippet name.
